File: Assets/Lua/Launch/Command/Impl/CommandSelector.lua
require 'Launch/Command/CommandContainerBase'
--选择执行器
CommandSelector = Class(CommandContainerBase)
function CommandSelector:ctor()
end

function CommandSelector:AddCondition(select,success,fail)
	self.select = select
	self.success = success
	self.fail = fail
	self.select.parent = self
	if(self.success ~= nil) then
		self.success.parent = self
	end
	if(self.fail ~= nil) then
		self.fail.parent = self
	end
end

function CommandSelector:OnChildDone(cmd)
	if(cmd == self.select) then
		if(self.select.state == CmdExecuteState.Success) then
			if(self.success ~= nil) then
				self.success:Execute(self.context)
			else
				self:OnExecuteDone(CmdExecuteState.Success)
			end
		else
			if(self.fail ~= nil) then
				self.fail:Execute(self.context)
			else
				self:OnExecuteDone(CmdExecuteState.Success)
			end
		end
	else
		CommandSelector.superclass.OnExecuteDone(self,cmd.state)
	end
end

function CommandSelector:Execute(context)
	CommandSelector.superclass.Execute(self,context)
	self.select:Execute(self.context)
end


function CommandSelector:OnDestroy()
	CommandSelector.superclass.OnDestroy(self)
	if(self.success ~= nil) then
		self:OnChildDestroy(self.success)
	end
	if(self.fail ~= nil) then
		self:OnChildDestroy(self.fail)
	end
end

File: Assets/Scripts/Framework/Command/Implement/CommandSelector.cs
using System;

namespace Launch
{
	public class CommandSelector : CommandContainerBase
	{
		protected CommandBase _selectWork;
		protected CommandBase _succeedWork;
		protected CommandBase _failWork;

		public void AddConditionWork(CommandBase selectWork,CommandBase succeedWork,CommandBase failWork)
		{
			_selectWork = selectWork;
			_succeedWork = succeedWork;
			_failWork = failWork;

			_selectWork.Parent = this;
			if (_succeedWork != null)
			{
				_succeedWork.Parent = this;
			}
			if (_failWork != null)
			{
				_failWork.Parent = this;
			}
		}

        public override void OnChildDone (CommandBase child)
		{
			if (child == this._selectWork) 
			{
				if (this._selectWork.State == CmdExecuteState.Success) 
				{
					if (_succeedWork != null)
					{
						_succeedWork.Execute (_context);
					}
					else
					{
						this.OnExecuteDone (CmdExecuteState.Success);
					}
				} 
				else 
				{
					if (_failWork != null)
					{
						_failWork.Execute (_context);
					}
					else
					{
                        this.OnExecuteDone(CmdExecuteState.Success);
                    }
				}
			} 
			else 
			{
                this.OnExecuteDone(child.State);
			}
		}

        public override void Execute(ICommandContext context)
        {
            base.Execute(context);
            _selectWork.Execute(context);
        }

		public override void OnDestroy ()
		{
			base.OnDestroy ();
			if (_succeedWork != null)
			{
				_succeedWork.OnDestroy ();
			}
			if (_failWork != null)
			{
				_failWork.OnDestroy ();
			}
		}
	}
}

